packages/framework/src/Model/Product/View/ListedProductViewFactory.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\FrameworkBundle\Model\Product\View;
+
+use Shopsys\FrameworkBundle\Component\Normalizer\Normalizer;
+use Shopsys\FrameworkBundle\Model\Product\Product;
+use Shopsys\FrameworkBundle\Component\Image\View\ImageViewFactory;
+
+/**
+ * @experimental
+ */
+class ListedProductViewFactory
+{
+    /**
+     * @var \Shopsys\FrameworkBundle\Component\Image\View\ImageViewFactory
+     */
+    protected $imageViewFactory;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\View\ProductActionViewFactory
+     */
+    protected $productActionViewsFactory;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Component\Normalizer\Normalizer
+     */
+    protected $normalizer;
+
+    /**
+     * @param \Shopsys\FrameworkBundle\Component\Image\View\ImageViewFactory $imageViewFactory
+     * @param \Shopsys\FrameworkBundle\Model\Product\View\ProductActionViewFactory $productActionViewsFactory
+     * @param \Shopsys\FrameworkBundle\Component\Normalizer\Normalizer $normalizer
+     */
+    public function __construct(
+        ImageViewFactory $imageViewFactory,
+        ProductActionViewFactory $productActionViewsFactory,
+        Normalizer $normalizer
+    ) {
+        $this->imageViewFactory = $imageViewFactory;
+        $this->productActionViewsFactory = $productActionViewsFactory;
+        $this->normalizer = $normalizer;
+    }
+
+    /**
+     * @param array $listedProductViewsData
+     * @return \Shopsys\FrameworkBundle\Model\Product\View\ListedProductView[]
+     */
+    public function getListedProductViews(array $listedProductViewsData): array
+    {
+        $imageViewsIndexedByProductIds = $this->imageViewFactory->getImageViewsOrNullsIndexedByEntityIds(
+            Product::class,
+            $this->getAllIds($listedProductViewsData)
+        );
+        $actionViewsIndexedByProductIds = $this->productActionViewsFactory->getProductActionViewsIndexedByProductIds($listedProductViewsData);
+
+        foreach ($listedProductViewsData as $index => $listedProductViewData) {
+            $productId = $listedProductViewData['id'];
+            $listedProductViewsData[$index]['action'] = $actionViewsIndexedByProductIds[$productId];
+            $listedProductViewsData[$index]['image'] = $imageViewsIndexedByProductIds[$productId];
+        }
+
+        return $this->normalizer->denormalizeArray($listedProductViewsData, $this->getViewObjectName());
```

Comment:
What is the benefit of denormalizing array of objects instead of object by object? Is there any?

Because it is a bit complicated for me - you have to prepare array of arrays, then pass it, the normalizer have to deal with array of arrays. For me it would be more readable denormalizing one object at a time...

```php
$result = [];
        foreach ($listedProductViewsData as $listedProductViewData) {
            $productId = $listedProductViewData['id'];
            $listedProductViewData['action'] = $actionViewsIndexedByProductIds[$productId];
            $listedProductViewData['image'] = $imageViewsIndexedByProductIds[$productId];
            
            $result[] = $this->normalizer->denormalize($listedProductViewData, $this->getViewObjectName());
        }

        return $result;
```

As you can see, I'm not a big fan of nested arrays

Reply:
I realized the code can be written in non magic fashion.... So I suggest even better

```php

    /**
     * @param array $listedProductViewsData
     * @return \Shopsys\FrameworkBundle\Model\Product\View\ListedProductView[]
     */
    public function getListedProductViews(array $listedProductViewsData): array
    {
        $enrichedData = $this->enrichData($listedProductViewsData);
        return array_map(
            function (array $data) {
                return $this->createView($data);
            },
            $enrichedData
        );
    }
    
    protected function enrichData(array $listedProductViewsData): array
    {
        $imageViewsIndexedByProductIds = $this->imageViewFactory->getImageViewsOrNullsIndexedByEntityIds(
            Product::class,
            $this->getAllIds($listedProductViewsData)
        );
        $actionViewsIndexedByProductIds = $this->productActionViewsFactory->getProductActionViewsIndexedByProductIds($listedProductViewsData);

        foreach ($listedProductViewsData as $index => $listedProductViewData) {
            $productId = $listedProductViewData['id'];
            $listedProductViewsData[$index]['action'] = $actionViewsIndexedByProductIds[$productId];
            $listedProductViewsData[$index]['image'] = $imageViewsIndexedByProductIds[$productId];
        }

        return $listedProductViewsData;
    }
    
    protected function createView(array $listedProductViewsData)
    {
        return new ListedProductView(
            $listedProductViewsData['id'],
            $listedProductViewsData['name'],
            $listedProductViewsData['image'],
            // ...
        );
    }
```